Libraries/LLM/Models/Llama.swift

```
@@ -51,39 +51,81 @@ func computeBaseFrequency(
-            base *= adjustedScale
+    private func computeFreqs() {
+        if ropeType != "llama3" {
+            freqs = nil
+            return
+        }
+
+        guard let ropeScaling = ropeScaling,
+            case .float(let factor) = ropeScaling["factor"],
+            case .float(let lowFreqFactor) = ropeScaling["low_freq_factor"] ?? .float(1.0),
+            case .float(let highFreqFactor) = ropeScaling["high_freq_factor"] ?? .float(4.0),
+            case .float(let oldContextLen) = ropeScaling["original_max_position_embeddings"]
+                ?? .float(8192)
+        else {
+            freqs = nil
+            return
         }
-        return MLXFast.RoPE(
-            x, dimensions: dims, traditional: traditional, base: base, scale: scale, offset: offset)
+
+        let lowFreqWavelen = oldContextLen / lowFreqFactor
+        let highFreqWavelen = oldContextLen / highFreqFactor
+
+        let indices = MLXArray(stride(from: 0, to: dims, by: 2))
+        var frequencies = MLX.pow(base, indices / Float(dims))
+        let wavelens = 2 * Float.pi * frequencies
+
+        frequencies = MLX.where(
+            wavelens .> MLXArray(lowFreqWavelen), frequencies * factor, frequencies)
+        let isMediumFreq = MLX.logicalAnd(
+            wavelens .> MLXArray(highFreqWavelen),
+            wavelens .< MLXArray(lowFreqWavelen)
+        )
+        let smoothFactors =
+            (oldContextLen / wavelens - lowFreqFactor) / (highFreqFactor - lowFreqFactor)
+        let smoothFreqs = frequencies / ((1 - smoothFactors) / factor + smoothFactors)
+
+        freqs = MLX.where(isMediumFreq, smoothFreqs, frequencies)
+    }
+
+    func callAsFunction(_ x: MLXArray, offset: Int = 0) -> MLXArray {
+        MLXFast.RoPE(
+            x,
+            dimensions: dims,
+            traditional: traditional,
+            base: base,
+            scale: scale,
+            offset: offset,
+            freqs: freqs
```

It should raise an error if both `base` and `freqs` are not `nil`. I'm wondering why it's not? Did you test it for when `freqs` is set? Am I missing somewhere that `base` gets a nil value?

---

Thanks. I think I've fixed this now.